### DISCUSSION TITLE
aws-cdk CLI - strange order of stack deletion

### DISCUSSION THREAD

**Discussion by antonostrovsky** [aws/aws-cdk#31699] [Q&A]
Hi CDK gurus!

We have encountered a strange condition where the order of stack declarations in CDK code appears to control stack destruction order. Example:
```
const stagingPdxEc2Stack = new Ec2Stack(app);
const stagingPdxLoadBalancerStack = new LoadBalancerStack(app);
const stagingRoute53Stack = new Route53Stack(app);
```
When I issue `bb cdk destroy stagingPdxEc2Stack` I get:
> Including depending stacks: stagingPdxLoadBalancerStack, stagingRoute53Stack
Are you sure you want to delete: stagingPdxLoadBalancerStack, stagingRoute53Stack, stagingPdxEc2Stack (y/n)? 

and when responding "Yes" we get:
> Failed to destroy stagingPdxLoadBalancerStack: UPDATE_COMPLETE (Export stagingPdxLoadBalancerStack:ExportsOutputFnGetAttstagingPdxLoadBalancerStackalbBEB5D449DNSName87411D04 cannot be deleted as it is in use by stagingRoute53Stack) 

However! When I change the order of declarations to:
```
const stagingRoute53Stack = new Route53Stack(app);  // <- just moved this up
const stagingPdxEc2Stack = new Ec2Stack(app);
const stagingPdxLoadBalancerStack = new LoadBalancerStack(app);
```
the **output of**  `bb cdk destroy stagingPdxEc2Stack`changes :
> Are you sure you want to delete: stagingPdxLoadBalancerStack, stagingPdxEc2Stack , stagingRoute53Stack

And now the deletion happens without the dependency problem.

We know of the existence of .addDependency, but we just really want someone to confirm that the order of stack declarations in the code affects the order of deployment/destruction. I've found [this commit](https://github.com/aws/aws-cdk/commit/29e904c2ff4f0719a3069144e979059b7d6c1f71) that controls this in aws-cdk, but I am not sure if this behaviour is normal.

Please advise, and thank you in advance!

**Comment by khushail** [comment]
Hi @antonostrovsky , AFAIK, the order of stack deletion is determined by the dependency between stacks , not the order of declaration. CloudFormation first deletes the stacks that do not have any dependencies on other stacks. Then it deletes the stacks that depend on the stacks that were just deleted, and so on until all stacks are deleted.

**Reply by antonostrovsky** [reply]
You will be surprised if you try the above. The order of declaration does matter in this case and it has already raised some eyebrows in the office :) 
I'm glad I've raised this as it felt like an unexpected behaviour, although perfectly reasonable (we should use .addDependency to help CDK figure out less obvious ones).
